### PR TITLE
Symbolical differentiation using TensorFlow

### PR DESCRIPTION
**Pull request by MatteoRobbiati**
Add symbolical differentiation to `operations.differentiation` collection.

### CODE REVIEW

**Review by alecandido — COMMENTED** [review]
Partial review

File: src/qiboml/operations/differentiation.py [hunk 1]
```diff
@@ -1,17 +1,21 @@
...
 
-def parameter_shift(
-    hamiltonian,
-    circuit,
-    parameter_index,
-    initial_state=None,
-    scale_factor=1,
-    nshots=None,
-    backend="qibojit",
+def one_parameter_shift(
```
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply src/qiboml/operations/differentiation.py:11]: I think this is partially true. 

1. it can be useful (in some situations) to compute the gradient w.r.t. a specific parameter only. 
2. ofc one can extract one gradient from the gradients list, but, if this is basically _for free_ using symbolical differentiation, in the case of using other rules (e.g. the parameter shift), the computational cost is drastically different.

**Review by alecandido — COMMENTED** [review]
Complete review

File: src/qiboml/operations/differentiation.py [hunk 2]
```diff
@@ -1,17 +1,21 @@
...
+    hamiltonian: qibo.hamiltonians.Hamiltonian,
+    circuit: qibo.Circuit,
+    parameter_index: int,
+    initial_state: Optional[Union[np.ndarray, qibo.Circuit]] = None,
+    scale_factor: float = 1.0,
```
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply src/qiboml/operations/differentiation.py:16]: Again partially true: scale factor takes into account the fact that one may want the derivative of `partial` of the angle.  E.g. an angle which is defined as $\theta' = \theta\cdot a + b$. Thus it cannot be tackled as a general scale factor, but as something which is related to the specific angle considered. 
Anyway, we should consider to use Mike's `Parameter` class here. That would be a nice improvement.

File: src/qiboml/operations/differentiation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alecandido) [reply src/qiboml/operations/differentiation.py:11]: Yes, but as I wrote, the way to go in that case would be to set a single trainable parameter.

Maybe it's not the most comfortable approach (though we can make a helper function that takes a circuit, the parameter specification, and return another circuit, being the exact same but with that single parameter marked as *trainable*), but there is no computational overhead, giving you access to the exact same computation.
The great advantage is to have a single uniform interface, and this will be highly beneficial.

File: src/qiboml/operations/differentiation.py [hunk 3]
```diff
@@ -158,3 +160,53 @@ def circuit(nqubits = 1):
...
+def parameter_shift(
+    hamiltonian: qibo.hamiltonians.Hamiltonian,
+    circuit: qibo.Circuit,
+    initial_state: Optional[Union[np.ndarray, qibo.Circuit]] = None,
+    nshots: int = None,
+    exec_backend: qibo.backends.Backend = NumbaBackend(),
+):
+    """Perform the parameter shift rule for all the parameters of a circuit."""
+    nparams = len(circuit.get_parameters())
+    gradients = []
```
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply src/qiboml/operations/differentiation.py:174]: Right, thanks!

File: src/qiboml/operations/differentiation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply src/qiboml/operations/differentiation.py:11]: Well well well, I got the point now. I agree, thanks for the suggestion!

File: src/qiboml/operations/differentiation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alecandido) [reply src/qiboml/operations/differentiation.py:16]: This you can also do after the gradient being returned. And I'd avoid any further structure on top of that, especially the `Parameter` class.

Moreover, I'd really avoid addressing something like $\theta' = \theta\cdot a + b$ inside the derivation: this is fully against composition.
Even TensorFlow, Pytorch, JAX, and whatever other differentiation framework are composable: they have rules for computing the gradients of building blocks, and one for the composition, the *chain rule*. And this is the basis of all autodiff, because you can automatically compose derivatives.

Thus, obtain the gradient w.r.t. the gate parameters as they are defined. If the parameters are functions of other parameters, the gradients should be composed (layers could come before and even after a circuit).

File: pyproject.toml [hunk 4]
```diff
@@ -14,6 +14,7 @@ packages = [{ include = "qiboml", from = "src" }]
 python = ">=3.9,<3.12"
 numpy = "^1.26.4"
 numba = "^0.59.0"
+qibojit = "^0.1.3"
```
[the remark this answers is not in the document]
Comment (alecandido) [reply pyproject.toml:17]: Already noted in https://github.com/qiboteam/qiboml/pull/19#discussion_r1601561225, I believe we're going to drop the dependency.

File: src/qiboml/operations/differentiation.py [hunk 5]
```diff
@@ -152,7 +82,102 @@ def circuit(nqubits = 1):
...
+                initial_state=initial_state,
+                nshots=nshots,
+                exec_backend=exec_backend,
+            )
+        )
+
+    return gradient
+
+
+def symbolical(
```
[the remark this answers is not in the document]
Comment (alecandido) [reply src/qiboml/operations/differentiation.py:140]: I had a look at the `tf.autodiff` module, but it seems that the only thing other than `GradientTape` is a [`ForwardAccumulator`](https://www.tensorflow.org/api_docs/python/tf/autodiff/ForwardAccumulator) (so, unexpectedly, TensorFlow can also differentiate in forward mode, though it is not advertising it that much...).

However, it seems that the only way to trace the graph to compute the gradient is evaluating the function with a not-so-dummy value (i.e. an actual value), that is used at the same time as a placeholder and as the value to compute the actual function.

I haven't found the way to decouple the two things.

The only thing we could save is the `circuit.get_parameters()`, that could all replaced with 0s. But we need to know how many of them...

[the remark this answers is not in the document]
Comment (alecandido) [reply src/qiboml/operations/differentiation.py:140]: > The only thing we could save is the `circuit.get_parameters()`, that could all replaced with 0s. But we need to know how many of them... (and possibly the whole layout, since it's using a list of tuples...)

I hope I'll be able to improve on this in `qibo-core`, https://github.com/qiboteam/qibo-core/issues/22.
But I'm just advertising that issue to trigger your feedback, the benefit of not using the actual parameters in this case is mostly conceptual. It would be nice to avoid doing the work twice, but we also want a TensorFlow independent way to evaluate the function in `expectation`.
So, for the time being, we are paying the price of evaluating twice...

Maybe, in this case, JAX/autograd is truly superior. Though it depends on the evaluation overhead, of course.

[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply src/qiboml/operations/differentiation.py:120]: I just added the jax version of this function. Even if it is ideally better, right now TF is faster.

File: src/qiboml/operations/differentiation.py [hunk 6]
```diff
@@ -152,7 +82,102 @@ def circuit(nqubits = 1):
...
+            circuits are executed.
+        initial_state (ndarray, optional): initial state on which the circuit
+            acts. Default is ``None``.
+
+    Returns:
+        (float): Gradient of the expectation value of the hamiltonian
+            with respect to circuit's variational parameters.
+
+    """
+    import tensorflow as tf  # pylint: disable=import-error
```
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply src/qiboml/operations/differentiation.py:148]: With the new push, I think this import should be fine.

File: src/qiboml/operations/differentiation.py [hunk 7]
```diff
@@ -152,7 +67,109 @@ def circuit(nqubits = 1):
...
+def symbolical_with_jax(
+    hamiltonian: qibo.hamiltonians.Hamiltonian,
+    circuit: qibo.Circuit,
+    exec_backend: qibo.backends.Backend,
+    initial_state: Optional[Union[np.ndarray, qibo.Circuit]] = None,
+):
+    def _expectation(params):
+        params = jax.numpy.array(params)
+        circuit.set_parameters(params)
+        return expectation._exact(hamiltonian, circuit, initial_state, exec_backend)
```
Comment (alecandido) [inline src/qiboml/operations/differentiation.py:171]: This is the function I was telling you yesterday: if this was already available, you didn't have to write it here, but just pass this function as input, and then this whole function would be just `jax.grad()`, plus the manipulations you need to make it compatible.

I.e. all the inputs of `symbolical_with_jax` are used in a single place, so the only input you need is the result of the operation in that place.

File: src/qiboml/operations/differentiation.py [hunk 8]
```diff
@@ -152,7 +67,109 @@ def circuit(nqubits = 1):
...
+
+    return jax.numpy.array(
+        [g[0].item() for g in jax.grad(_expectation)(circuit.get_parameters())]
```
Comment (alecandido) [inline src/qiboml/operations/differentiation.py:174]: Why do you need all this gymnastics with `[0].item()`? If you know

Comment (MatteoRobbiati) [reply src/qiboml/operations/differentiation.py:174]: It's the way I managed to make it work compatibly with the expectation interface. Maybe some easier solution can be found.

File: src/qiboml/operations/differentiation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alecandido) [reply src/qiboml/operations/differentiation.py:148]: Sorry, I'm not sure why, even because I'm not sure to what push you're referring to.

File: src/qiboml/operations/differentiation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alecandido) [reply src/qiboml/operations/differentiation.py:120]: It's quite possible that, even ideally, the speedup is rather limited. Evaluating the function could be cheaper than evaluating the gradient (because the gradient has an extra dimension on the derivation variables).

Still, it's pretty much unexpected that one is slower than the other...

Are you comparing TF-derivation-with-TF-backend, i.e. `(TF, sym, TF)`, with `(JAX, sym, JAX)`?
It could be that the slowdown is just the same of the JAX backend execution, and nothing new.

File: src/qiboml/operations/differentiation.py (hunk 8 above)
Comment (alecandido) [reply src/qiboml/operations/differentiation.py:174]: But why `jax.grad(_expectation)(circuit.get_parameters())` is incompatible?

File: pyproject.toml [hunk 9]
```diff
@@ -34,6 +34,8 @@ pylint = "^2.17"
 pytest-cov = "^3.0.0"
 pytest-env = "^0.8.1"
 pytest-benchmark = { version = "^4.0.0", extras = ["histogram"] }
+qibojit = "^0.1.3"
```
Comment (alecandido) [inline pyproject.toml:37]: Why is qibojit here again?

Comment (MatteoRobbiati) [reply pyproject.toml]: Thanks. Removed both `qibojit` and `numba`.
